network/src/connectivity_manager/mod.rs

```diff
@@ -435,22 +435,45 @@ where
     }
 
     fn choose_peers_to_dial(&mut self) -> Vec<(PeerId, DiscoveredPeer)> {
-        let to_connect: Vec<_> = self
+        let network_id = self.network_context.network_id();
+        let role = self.network_context.role();
+        let roles_to_dial = network_id.upstream_roles(&role);
+        let mut eligible: Vec<_> = self
             .discovered_peers
             .0
             .iter()
             .filter(|(peer_id, peer)| {
                 peer.is_eligible_to_be_dialed() // The node is eligible to dial
-                    && self.connected.get(peer_id).is_none() // The node is not already connected.
-                    && self.dial_queue.get(peer_id).is_none() // There is no pending dial to this node.
+                && !self.connected.contains_key(peer_id) // The node is not already connected.
+                && !self.dial_queue.contains_key(peer_id) // There is no pending dial to this node.
+                && roles_to_dial.contains(&peer.role) // We can dial this role
             })
             .collect();
 
+        // Prioritize by PeerRole
+        let rng = &mut self.rng;
+        eligible.sort_by(|(_, peer), (_, other)| {
+            let order = peer.role.cmp(&other.role);
+
+            // Randomly mix equal rank peers to ensure we don't get stuck on certain peers
+            if Ordering::Equal == order {
+                if rng.gen_bool(0.5) {
+                    Ordering::Less
+                } else {
+                    Ordering::Greater
+                }
```

Comment:
considering this is dynamic it seems ... scary ... should we instead consider weighting peers beforehand? Namely, we could imagine each tier is a whole integer and then we use randomization [0, 1) to weight them? Then we sort.

Reply:
You described basically what I had before I switched this this sorting (which is also why I didn't like this mode).  Originally, I had it sorted by levels, and then choose random ones at each level.  The only reason for the randomization anyways is that we don't have smarter information about which discovered peers work and which ones don't, meaning connections could be blocked by some bad nodes.

Reply:
If you `eligible.shuffle` then `eligible.sort_by_unstable` you'll sample randomly within each level

Reply:
Shuffled then sorted.